pkg/gadget-collection/gadgets/top/ebpf/gadget.go

```diff
@@ -0,0 +1,273 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpf
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cilium/ebpf"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/bpfstats"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadget-collection/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/ebpftop/piditer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/ebpftop/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+type Trace struct {
+	resolver gadgets.Resolver
+
+	started  bool
+	donechan chan bool
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+}
+
+func NewFactory() gadgets.TraceFactory {
+	return &TraceFactory{
+		BaseFactory: gadgets.BaseFactory{DeleteTrace: deleteTrace},
+	}
+}
+
+func (f *TraceFactory) Description() string {
+	t := `ebpftop shows cpu time used by ebpf programs.
+
+The following parameters are supported:
+ - %s: Output interval, in seconds. (default %d)
+ - %s: Maximum rows to print. (default %d)
+ - %s: The field to sort the results by (%s). (default %s)`
+	return fmt.Sprintf(t, types.IntervalParam, types.IntervalDefault,
+		types.MaxRowsParam, types.MaxRowsDefault,
+		types.SortByParam, strings.Join(types.SortBySlice, ","), types.SortByDefault)
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	trace := t.(*Trace)
+	if trace.started {
+		trace.stop()
+	}
+}
+
+func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
+	n := func() interface{} {
+		return &Trace{
+			resolver: f.Resolver,
+		}
+	}
+
+	return map[string]gadgets.TraceOperation{
+		"start": {
+			Doc: "Start ebpftop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Start(trace)
+			},
+		},
+		"stop": {
+			Doc: "Stop ebpftop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Stop(trace)
+			},
+		},
+	}
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		trace.Status.State = "Started"
+		return
+	}
+
+	t.donechan = make(chan bool)
+	traceName := gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+
+	// Enable stats collection
+	err := bpfstats.EnableBPFStats()
+	if err != nil {
+		trace.Status.OperationError = err.Error()
+		return
+	}
+
+	useFallbackIterator := false
+
+	// To resolve pids, we will first try to iterate using a bpf
+	// program. If that doesn't work, we will fall back to scanning
+	// all used fds in all processes /proc/$pid/fdinfo/$fd.
+	iter, err := piditer.NewPidIter()
+	if err != nil {
+		useFallbackIterator = true
+	}
+
+	t.started = true
+
+	go func() {
```

Comment:
Since this go function is 120 lines long, it could be moved in a separate function.
```
go t.run()
```

This will require moving `useFallbackIterator` and `iter` as fields of the `Trace struct`.